rama-tls-acme/src/client.rs

```diff
@@ -0,0 +1,664 @@
+use crate::proto::common::{ProtectedHeaderAcme, ProtectedHeaderCrypto, ProtectedHeaderKey};
+
+use super::proto::{
+    client::FinalizePayload,
+    client::{CreateAccountOptions, KeyAuthorization, NewOrderPayload},
+    common::{self},
+    server::REPLAY_NONCE_HEADER,
+    server::{self, Problem},
+};
+
+use base64::{Engine as _, prelude::BASE64_URL_SAFE_NO_PAD};
+use parking_lot::Mutex;
+use rama_core::{
+    Context,
+    error::{ErrorContext, ErrorExt, OpaqueError},
+    service::BoxService,
+};
+use rama_crypto::{
+    dep::{
+        pki_types::PrivatePkcs8KeyDer,
+        rcgen::{self, Certificate},
+    },
+    jose::{EMPTY_PAYLOAD, EcdsaKey, Empty, Headers, JWSBuilder, NO_PAYLOAD, Signer},
+};
+use rama_http::{
+    BodyExtractExt, Request, Response,
+    dep::http_body_util::BodyExt,
+    headers::{ContentType, HeaderMapExt, Location, RetryAfter, TypedHeader, UserAgent},
+    service::client::HttpClientExt,
+    utils::HeaderValueGetter,
+};
+use serde::Serialize;
+use std::{
+    borrow::Cow,
+    time::{Duration, SystemTime},
+};
+use tokio::time::{sleep, timeout};
+
+/// Acme client that will used for all acme operations
+pub struct AcmeClient {
+    https_client: AcmeHttpsClient,
+    directory: server::Directory,
+    nonce: Mutex<Option<String>>,
+}
+
+/// Alias for http client used by acme client
+pub type AcmeHttpsClient = BoxService<(), Request, Response, OpaqueError>;
+
+impl AcmeClient {
+    /// Create a new acme [`AcmeClient`] for the given directory url and using the provided https client
+    pub async fn new(
+        directory_url: &str,
+        https_client: AcmeHttpsClient,
```

Comment:
I would just accept a `Service` and box it yourself :) the `into_boxed` is a nop operation if it's already a boxed service